### ISSUE TITLE
Branch pruning not recognizing fact that a type is never None (original: Input type None is not recognized)

### ISSUE THREAD

**Issue by ogauthe** [numba/numba#8603]
Hello,

I want to write a function that is basically summing two arrays. One of them may be missing and is replaced by `None`. They cannot be both `None`. As far as I understand numba, 3 different functions should be compiled in the 3 different cases, as `None` is a different type from array. I also expect the compiler to remove the dead parts of the code depending on the input types. However I get an error with this function:

```python
import numpy as np
import numba

@numba.njit
def f(a, b): 
    if a is not None:
        c = np.zeros((1,), dtype=a.dtype)
        c += a   
    else:    
        c = np.zeros((1,), dtype=b.dtype)
    if b is not None:
        c += b   
    return c 

x = np.ones((1,))
y = np.ones((1,))
print(f(x, y))
print(f(None, y))
print(f(x, None))
```

```
[2.]
[1.]
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
Input In [49], in <cell line: 19>()
     17 print(f(x, y))
     18 print(f(None, y))
---> 19 print(f(x, None))

File ~/miniconda3/lib/python3.9/site-packages/numba/core/dispatcher.py:468, in _DispatcherBase._compile_for_args(self, *args, **kws)
    464         msg = (f"{str(e).rstrip()} \n\nThis error may have been caused "
    465                f"by the following argument(s):\n{args_str}\n")
    466         e.patch_message(msg)
--> 468     error_rewrite(e, 'typing')
    469 except errors.UnsupportedError as e:
    470     # Something unsupported is present in the user code, add help info
    471     error_rewrite(e, 'unsupported_error')

File ~/miniconda3/lib/python3.9/site-packages/numba/core/dispatcher.py:409, in _DispatcherBase._compile_for_args.<locals>.error_rewrite(e, issue_type)
    407     raise e
    408 else:
--> 409     raise e.with_traceback(None)

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'dtype' of type none

File "<ipython-input-49-2844d1d2e619>", line 10:
def f(a, b): 
    <source elided>
    else:    
        c = np.zeros((1,), dtype=b.dtype)
        ^

During: typing of get attribute at <ipython-input-49-2844d1d2e619> (10)

File "<ipython-input-49-2844d1d2e619>", line 10:
def f(a, b): 
    <source elided>
    else:    
        c = np.zeros((1,), dtype=b.dtype)
        ^
```

The error is always for `f(x, None)`, even if I remove `f(x, y)` or `f(None, y)`. There is no error when I remove `@njit`.

- [X] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [X] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

numpy version `1.21.5`
numba version `0.55.1`

**Comment by sklam** [comment]
Thanks for reporting the issue.

**TL;DR**: Numba is not pruning branches base on the fact that a value can never be `None`.

Debugging details:

Numba's error message is leaving some hints of what it's thinking:

```pytb
Unknown attribute 'dtype' of type none

File "<ipython-input-49-2844d1d2e619>", line 10:
def f(a, b): 
    <source elided>
    else:    
        c = np.zeros((1,), dtype=b.dtype)
        ^

During: typing of get attribute at <ipython-input-49-2844d1d2e619> (10)
```

It seems Numba doesn't prune branches when the argument value cannot be `None`. It has preserved first if-else even though it can know that `a` is not `None`.

(ping @stuartarchibald )

**Comment by sklam** [comment]
Categorizing this as a feature-request on enhancing the dead-branch-pruner to handle cases when a value cannot be `None`.

**Comment by sklam** [comment]
For a workaround...

since numba can only deal with facts that a value is `None`, a guard can be inserted to the `else` branch either as:

```python

    else:
        assert b is not None
        c = np.zeros((1,), dtype=b.dtype)
```
or

```python

    elif b is not None:
        c = np.zeros((1,), dtype=b.dtype)
```
